### DISCUSSION TITLE
Hide accessories

### DISCUSSION THREAD

**Discussion by rayr007** [thoukydides/homebridge-homeconnect#104] [Q&A] [answered]
Is there a way to hide some of the homeconnect appliances from homekit? Or make an appliance read-only? The issue I've run into is siri being stupid and missing the work "lights" when my spouse says "hey siri, turn off kitchen lights". Siri decides to turn off kitchen which includes stopping the dishwasher that is running. In the past, I believe the fridge was also powered off but perhaps power for the fridge is no longer mapped by default?

Anyway, I don't mind notifications of when the dishwasher is done. Ideally, I'd like to disable homekit being able to stop it. But I'd compromise and skip that appliance completely if I had to.

**Comment by thoukydides** [comment]
> Is there a way to hide some of the homeconnect appliances from homekit?

No. This plugin exposes all of the appliances that are associated with the Home Connect account.

Hiding specific appliances would require a significant restructuring of the plugin to avoid breaking the configuration schema for homebridge-config-ui-x. I have no plans to do that.

> Or make an appliance read-only?

Again, no.

This would actually be more complicated to implement than hiding the whole appliance, since the plugin needs to control the appliance to reliably read its capabilities in order to expose the appropriate functionality to HomeKit and to generate the homebridge-config-ui-x configuration schema.

> The issue I've run into is siri being stupid and missing the work "lights" when my spouse says "hey siri, turn off kitchen lights". Siri decides to turn off kitchen which includes stopping the dishwasher that is running.

Yes, I have noticed the same on occasions. It is very annoying. It's a shame that it is so difficult to build common sense into "smart" assistants...

> In the past, I believe the fridge was also powered off but perhaps power for the fridge is no longer mapped by default?

That was definitely not the case. The Home Connect API has never supported switching off the power to cooling appliances. At most it could have switched off super-cooling, eco mode, sabbath mode, vacation mode, and fresh modes. Normal fridge operation would have remained on.

> Anyway, I don't mind notifications of when the dishwasher is done. Ideally, I'd like to disable homekit being able to stop it. But I'd compromise and skip that appliance completely if I had to.

The only HomeKit notifications that I get with this plugin are for doors open/closed, following the change in v0.25.0. Those notifications can be disabled for specific appliances in the Home app (or in Settings on an Apple TV).

Are you sure that the dishwasher finished notifications aren't from the Home Connect app itself?

**Reply by urbanhaas** [reply]
Thanks for this answer, as I too was looking at the same issue. 

One of the ideas implemented by the homebridge-kumo devs was to have an option to simplify the accessory as one simplified control. For me I have a washer and dryer, and each possible setting shows as a switch, which is crazy. TBH, on HomeKit, I'm just looking for status and don't really want to see 16 switches for each accessory. 

Might a workaround, or an improvement, be to implement a simplified accessory option that places everything under one control?

**Comment by thoukydides** [comment]
@urbanhaas, the main problem is that HomeKit does not have an object model that is appropriate for kitchen appliances. This forces a choice between various sub-optimal options:
1. Only implement the minimal functionality that is directly supported by the HomeKit object model, which in the case of this plugin would be just the power switch (plus fan and light control for Hood appliances). This wouldn't make the integration with HomeKit useful.
2. Mis-use the HomeKit object model, mapping functionality to Services and Characteristics that have the required control mechanisms even if their semantics are entirely wrong. This generally breaks control via Siri.
3. Define proprietary Services and Characteristics for the functionality which does not map directly to the existing HomeKit model. These will not be shown by Apple's Home app and cannot be controlled by Siri.
4. Map functionality to generic HomeKit Services and Characteristics that have minimal semantics, such as Switch and Stateless Programmable Switch services.

My early attempts at HomeKit plugins tried options 2 and 3, but since I mainly use Siri as the method of controlling HomeKit I quickly abandoned those approaches. Hence, this plugin mainly uses option 4... which is why each appliance exposes its functionality mainly as a series of generic switches.

This plugin creates a single HomeKit Accessory for each Home Connect appliance. All of the associated Services and Characteristics are part of that single Accessory. You then have the choice of whether the Home app displays the Accessory as two tiles (one for the Stateless Programmable Switch services that can be used to trigger automations, and one for all the other Services) or displaying separate tiles for each Service. You can switch between these using the "Show as Single Tile" and "Show as Separate Tiles" options within the Accessory's configuration. That seems to be what you are requesting by "places everything under one control", and is already supported by this plugin.

The Kumo plugin isn't a particularly relevant example here since it looks like it only handles heating/cooling systems, which are directly supported by the HomeKit object model.

**Reply by urbanhaas** [reply]
@thoukydides, thank you for your reply. I understand your reply, but I might not agree.

I have two types of users in my home: tech savvy and hardly any tech. For the latter, a simple control is desirable, with as little tiles and switches within the tile. For the more tech savvy, they want to see everything, so they might write custom automations for all the use cases.

Currently, opening my washer or dryer, shows a myriad of state switches for each possible run program, which is selected by the dial on front. They are all mutually exclusive, only one program is selected on the dial at a time, so each showing as a switch is kind of insane even from a home automation programmability standpoint. All the switches are noise in this case:
![367BCD01-40A9-4544-8EB0-71059E669F31](https://user-images.githubusercontent.com/42039536/191991053-f35e6f2f-66a5-43fd-bb91-fa361c0590bd.png)

There is a different control which more closely matches what I'm trying to imagine it looking like, the TV with input selector:

![E93F40F3-8677-46B3-B6DB-8697CE149E04](https://user-images.githubusercontent.com/42039536/191991355-16a0ddab-c15e-4139-bce4-31776d118353.png)

This much more closely matches the characteristics of the washer and dryer (and likely other appliances). 

You're implementation has some really nice features, especially the push buttons for triggering automations on events. I'd love to see it cleaned up a bit. I'm a developer, but I haven't touched HomeBridge or any plugins. I might be willing to dive in.

You mentioned the new frameworks. I am an iOS developer, I do have iOS 16.1 beta. I understood the changes to have mostly Matter support, but I haven't really dived into HomeKit frameworks before. I might have to look and see how the frameworks may be changing under the hood with the Matter/Thread implementation landing next month.

**Reply by thoukydides** [reply]
This plugin already supports customisation of which programs to present as switches - all, none, or a custom set. This is fully described on https://github.com/thoukydides/homebridge-homeconnect/wiki/Programs

I do not consider presenting the appliance programs as a TV Inputs a good option since it will have the wrong semantics, and is likely to cause problems with Siri control - especially if another Accessory is using that Service type appropriately.

My expectation is that Apple will update the HomeKit object model to fully support Matter's object model. That may allow more appropriate mappings of Home Connect functionality, in which case I will update this plugin. Otherwise, I will be sticking with my current design approach since it matches my own usage.

If you dislike the approach I have taken then you always have the option to fork this plugin or write your own.

**Reply by thoukydides** [reply]
> You mentioned the new frameworks. I am an iOS developer, I do have iOS 16.1 beta. I understood the changes to have mostly Matter support, but I haven't really dived into HomeKit frameworks before. I might have to look and see how the frameworks may be changing under the hood with the Matter/Thread implementation landing next month.

The Matter 1.0 specifications have just been published:
https://csa-iot.org/newsroom/matter-arrives/
https://csa-iot.org/developer-resource/specifications-download-request/

Unfortunately, the device types and clusters that Matter defines are very similar to those already supported by HomeKit. It does not look like it will improve the support for kitchen appliances.

**Comment by rayr007** [comment]
I like how you allow putting in a list of "do not map" programs in the JSON. What I was originally looking for is the exact same sort of feature for the entire appliance. Essentially, the mapping doesn't work for me or has no value for this appliance so "do not map appliance" in the config. If I specify an appliance and give it an empty list of programs, perhaps that will achieve what I'm looking for?

**Comment by thoukydides** [comment, marked as answer]
Yes, I understand your original request. However, this plugin is built around having a 1:1 mapping between Home Connect appliances and HomeKit accessories. The configuration for a particular appliance/accessory is not checked until after the HomeKit accessory has been created.

Setting `"addprograms": "none"` is the closest that it is supported. However, you will still get switches created for any non-program settings that the appliance supports, such as child-lock, plus stateless programmable switches for events. There may also be light bulb controls if it supports a functional/ambient light, or a door if its status can be monitored. I do not have any plans to add configuration options to disable any of that.
